exporter/elasticsearchexporter/exporter.go

```diff
@@ -193,21 +191,18 @@ func (e *elasticsearchExporter) pushMetricsData(
 	}
 	defer session.End()
 
-	var (
-		validationErrs []error // log instead of returning these so that upstream does not retry
-		errs           []error
-	)
+	var errs []error
 	resourceMetrics := metrics.ResourceMetrics()
 	for i := 0; i < resourceMetrics.Len(); i++ {
 		resourceMetric := resourceMetrics.At(i)
 		resource := resourceMetric.Resource()
 		scopeMetrics := resourceMetric.ScopeMetrics()
 
-		resourceDocs := make(map[string]map[uint32]objmodel.Document)
-
 		for j := 0; j < scopeMetrics.Len(); j++ {
+			var validationErrs []error // log instead of returning these so that upstream does not retry
 			scopeMetrics := scopeMetrics.At(j)
 			scope := scopeMetrics.Scope()
+			groupedDataPointsByIndex := make(map[string]map[uint32][]dataPoint)
```

Comment:
Note to reviewer: I made it so that documents from different scopes are never merged. This simplified the serialization logic and also fixes a subtle bug in the current implementation where we're only hashing the scope attributes but not the scope name. This leads to grouping of potentially different scopes to the same document. I guess as a consequence, we should also add the scope name as a dimension in the mappings.

Reply:
I think by moving this here, rather than outside of the `scopeMetrics` loop, we're assuming that there will never be two identical scopes within a resource. Is that a safe assumption?

Reply:
I suppose it's no worse than the existing assumption that resourceMetrics is free of duplicate resources.

Reply:
What will make it safe is that the only consequence of being wrong in the assumption is leaving some storage savings on the table. In other words, we should prioritize https://github.com/elastic/elasticsearch/issues/99123, which turns out to be more of an issue than anticipated in various contexts.

Reply:
Wouldn't duplicates resources/scopes lead to duplicate _tsid & doc rejections?

Reply:
Yes, it would, until we fix the referenced issue.

Reply:
Fixing the underlying ES issue turns out to be very tricky. I've created a PR for the ES exporter to enable the grouping across resources and scopes, which hopefully lessens the impact of the bug: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37509.